### ISSUE TITLE
How to conver from `Box<Error>` into `failure::Error`?

### ISSUE THREAD

**Issue by matklad** [rust-lang-nursery/failure#158]
So, I have a legacy API which returns a `Box<::std::error::Error + Send + Sync + 'static>`, and I want to wrap this API in a shiny new `failure::Error`. Is there any existing compat type which can help me? Calling `.into()` on the boxed error does not help :( 

```
  --> explorer/mod.rs:79:49
   |
79 |                     let e: ::failure::Error = e.into();
   |                                                 ^^^^ `iron::Error + std::marker::Sync + std::marker::Send` does not have a constant size known at compile-time
   |
   = help: the trait `std::marker::Sized` is not implemented for `iron::Error + std::marker::Sync + std::marker::Send`
   = note: required because of the requirements on the impl of `iron::Error` for `std::boxed::Box<iron::Error + std::marker::Sync + std::marker::Send>`
   = note: required because of the requirements on the impl of `failure::Fail` for `std::boxed::Box<iron::Error + std::marker::Sync + std::marker::Send>`
   = note: required because of the requirements on the impl of `std::convert::From<std::boxed::Box<iron::Error + std::marker::Sync + std::marker::Send>>` for `failure::Error`
   = note: required because of the requirements on the impl of `std::convert::Into<failure::Error>` for `std::boxed::Box<iron::Error + std::marker::Sync + std::marker::Send>`
```

**Comment by withoutboats** [comment]
I think its not possible today, but I would like it to be. `failure::Error` is ultimately a wrapper around `Box<Fail>`. The problem is that there's no way to cast `dyn Error` to `dyn Fail`, even though the underlying concrete type implements `Fail` (because all `Error + Send + Sync + 'static` implement `Fail`). We don't have a mechanism for locating the vtable dynamically.

**Comment by withoutboats** [comment]
If there is a way to find the write vtable, or otherwise bridge this, I'd be very eager to include a constructor that takes a Box<Error>!

**Comment by matklad** [comment]
> or otherwise bridge this

Could we have a `BoxCompat(Box<::std::error::Error>): Fail` wrapper and an `Error::compat_from_box` constructor? Granted, it's not as fancy as `Into`, and the runtime indirection has an inelegant double indirection, but it's still better than nothing.

I think we can even hide `BoxCompat` struct, so that, if in the future we figure out `dyn` cast trick, it would be possible to change implementation?

**Comment by withoutboats** [comment]
Yes, that should work, and I'd like to hide it behind a single constructor on `Error` as you said, so that the intermediate type can be removed someday if possible.

**Comment by mitsuhiko** [comment]
This was fixed by #159
